File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressQuantumTreeList/Demos/CBuilder/CalculatedFieldsDemo/CalculatedFieldsDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "CalculatedFieldsDemoMain.h"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxClasses"
#pragma link "cxLookAndFeels"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDBTL"
#pragma link "cxEditRepositoryItems"
#pragma link "cxInplaceContainer"
#pragma link "cxMaskEdit"
#pragma link "cxTL"
#pragma link "cxTLData"
#pragma link "cxCheckBox"
#pragma link "cxButtons"
#pragma link "cxContainer"
#pragma link "cxCurrencyEdit"
#pragma link "cxGroupBox"
#pragma link "cxLabel"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxTLdxBarBuiltInMenu"
#pragma link "dxGallery"
#pragma link "dxGalleryControl"
#pragma link "dxmdaset"
#pragma link "dxSpreadSheetTypes"
#pragma link "dxSpreadSheetUtils"
#pragma resource "*.dfm"
TfrmMain *frmMain;
//---------------------------------------------------------------------------
__fastcall TfrmMain::TfrmMain(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::FormCreate(TObject *Sender)
{
  String APath = ExtractFilePath(Application->ExeName) + "..\\..\\Data\\";
  cdsProducts->LoadFromFile(APath + "Products2.cds");
  mdOrder->LoadFromBinaryFile(APath + "Order2TL.dat");
  ((TdxGalleryControlOptionsBehaviorAccess*)galColumns->OptionsBehavior)->ItemHotTrack = false;
  TreeList->FullExpand();
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::mdOrderCalcFields(TDataSet *DataSet)
{
  if (cdsProducts->FindKey(ARRAYOFCONST((mdOrder->FieldByName("ProductID")->AsInteger))))
	mdOrder->FieldByName("ProductName")->AsString = cdsProducts->FieldByName("ProductName")->AsString;
  else
	mdOrder->FieldByName("ProductName")->AsString = Format(AnsiString("Order #%d"), ARRAYOFCONST((mdOrder->FieldByName("OrderID")->AsInteger)));
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::btnShowExpressionEditorClick(TObject *Sender)
{
  if (galColumns->Gallery->Groups->Groups[0]->Items->Items[0]->Checked)
	tlcTotal->ShowExpressionEditor();
  else
	tlcAmountDiscount->ShowExpressionEditor();
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::TreeListEditing(TcxCustomTreeList *Sender, TcxTreeListColumn *AColumn,
          bool &Allow)
{
  Allow = (Allow && !(TreeList->FocusedNode->IsGroupNode));
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::TreeListStylesGetContentStyle(TcxCustomTreeList *Sender,
          TcxTreeListColumn *AColumn, TcxTreeListNode *ANode, TcxStyle *&AStyle)

{
  if (ANode->IsGroupNode)
        AStyle = stGroup;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::TreeListSummary(TcxCustomTreeList *ASender, const TcxTreeListSummaryEventArguments &Arguments,
          TcxTreeListSummaryEventOutArguments &OutArguments)

{
  OutArguments.Done = Arguments.Node->IsGroupNode;
}
//---------------------------------------------------------------------------


void __fastcall TfrmMain::ValidateExpressionColumn(TcxTreeListColumn *Sender,
          TcxTreeListNode *ANode, const Variant &AValue, TcxEditValidateInfo *AData)

{
  int AErrorCode = ANode->ErrorCodes[Sender->ItemIndex];
  if (AErrorCode > 0) {
	AData->ErrorType = eetError;
	AData->ErrorText = dxSpreadSheetErrorCodeToString((TdxSpreadSheetFormulaErrorCode)AErrorCode);
  }
}
//---------------------------------------------------------------------------

File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressQuantumTreeList/Demos/Delphi/CalculatedFieldsDemo/CalculatedFieldsDemoMain.pas
unit CalculatedFieldsDemoMain;

{$I cxVer.inc}

interface

uses
  SysUtils, Classes, Controls, Forms, Menus, DB, StdCtrls, ComCtrls, DBClient,
  dxCore, cxStyles, cxCustomData, cxGraphics, cxFilter, cxData, cxDataStorage,
  cxEdit, cxClasses, cxControls, cxEditRepositoryItems, cxLookAndFeels, cxLookAndFeelPainters,
  cxNavigator, cxContainer, cxGroupBox, ActnList, cxCheckBox, cxLabel,
  cxTextEdit, cxMaskEdit, cxSpinEdit, cxDropDownEdit, cxImageComboBox, cxDBData,
  cxGridDBTableView, DemoBasicMain, MidasLib, Dialogs, ImgList, cxTL,
  cxTLdxBarBuiltInMenu, cxDataControllerConditionalFormattingRulesManagerDialog,
  cxInplaceContainer, cxDBTL, cxTLData, dxmdaset, dxLayoutContainer, dxLayoutControl, cxCurrencyEdit, dxGallery,
  dxGalleryControl, dxLayoutControlAdapters, cxButtons, dxLayoutLookAndFeels;

type
  TfrmMain = class(TDemoBasicMainForm)
    cdsProducts: TClientDataSet;
    cdsProductsProductID: TIntegerField;
    cdsProductsProductName: TStringField;
    cdsProductsSupplierID: TIntegerField;
    cdsProductsCategoryID: TIntegerField;
    cdsProductsQuantityPerUnit: TStringField;
    cdsProductsUnitPrice: TFloatField;
    cdsProductsUnitsInStock: TIntegerField;
    cdsProductsUnitsOnOrder: TIntegerField;
    cdsProductsReorderLevel: TIntegerField;
    cdsProductsDiscontinued: TBooleanField;
    cdsProductsEAN13: TStringField;
    mdOrder: TdxMemData;
    IntegerField1: TIntegerField;
    mdOrderParentOrderID: TIntegerField;
    IntegerField2: TIntegerField;
    FloatField1: TFloatField;
    IntegerField3: TIntegerField;
    FloatField2: TFloatField;
    DateTimeField1: TDateTimeField;
    mdOrderProductName: TStringField;
    dsOrder: TDataSource;
    cxStyleRepository1: TcxStyleRepository;
    stGroup: TcxStyle;
    cxGroupBox1: TcxGroupBox;
    TreeList: TcxDBTreeList;
    tlcParentOrderID: TcxDBTreeListColumn;
    tlcOrderID: TcxDBTreeListColumn;
    tlcProductName: TcxDBTreeListColumn;
    tlcUnitPrice: TcxDBTreeListColumn;
    tlcQuantity: TcxDBTreeListColumn;
    tlcDiscount: TcxDBTreeListColumn;
    tlcAmountDiscount: TcxDBTreeListColumn;
    tlcTotal: TcxDBTreeListColumn;
    cxGroupBox2: TcxGroupBox;
    btnShowExpressionEditor: TcxButton;
    galColumns: TdxGalleryControl;
    dxGalleryControl1Group1: TdxGalleryControlGroup;
    dxGalleryControl1Group1Item1: TdxGalleryControlItem;
    dxGalleryControl1Group1Item2: TdxGalleryControlItem;
    cxLabel1: TcxLabel;
    procedure FormCreate(Sender: TObject);
    procedure mdOrderCalcFields(DataSet: TDataSet);
    procedure btnShowExpressionEditorClick(Sender: TObject);
    procedure TreeListEditing(Sender: TcxCustomTreeList; AColumn: TcxTreeListColumn; var Allow: Boolean);
    procedure TreeListStylesGetContentStyle(Sender: TcxCustomTreeList; AColumn: TcxTreeListColumn;
      ANode: TcxTreeListNode; var AStyle: TcxStyle);
    procedure TreeListSummary(ASender: TcxCustomTreeList; const Arguments: TcxTreeListSummaryEventArguments;
      var OutArguments: TcxTreeListSummaryEventOutArguments);
    procedure ValidateExpressionColumn(Sender: TcxTreeListColumn;
      ANode: TcxTreeListNode; const AValue: Variant;
      AData: TcxEditValidateInfo);
  end;

var
  frmMain: TfrmMain;

implementation

uses
  dxSpreadSheetTypes, dxSpreadSheetUtils, AboutDemoForm;

{$R *.dfm}

type
  TdxGalleryControlOptionsBehaviorAccess = class(TdxGalleryControlOptionsBehavior);

procedure TfrmMain.FormCreate(Sender: TObject);
var
  APath: string;
begin
  APath := ExtractFilePath(Application.ExeName) + '..\..\Data\';
  cdsProducts.LoadFromFile(APath + 'Products2.cds');
  mdOrder.LoadFromBinaryFile(APath + 'Order2TL.dat');
  TdxGalleryControlOptionsBehaviorAccess(galColumns.OptionsBehavior).ItemHotTrack := False;
  TreeList.FullExpand;
end;

procedure TfrmMain.mdOrderCalcFields(DataSet: TDataSet);
begin
  if cdsProducts.FindKey([mdOrder.FieldByName('ProductID').AsInteger]) then
    mdOrder.FieldByName('ProductName').AsString := cdsProducts.FieldByName('ProductName').AsString
  else
    mdOrder.FieldByName('ProductName').AsString := Format('Order #%d', [mdOrder.FieldByName('OrderID').AsInteger]);
end;

procedure TfrmMain.TreeListEditing(Sender: TcxCustomTreeList; AColumn: TcxTreeListColumn; var Allow: Boolean);
begin
  Allow := Allow and not TreeList.FocusedNode.IsGroupNode;
end;

procedure TfrmMain.TreeListStylesGetContentStyle(Sender: TcxCustomTreeList; AColumn: TcxTreeListColumn;
  ANode: TcxTreeListNode; var AStyle: TcxStyle);
begin
  if ANode.IsGroupNode then
    AStyle := stGroup
end;

procedure TfrmMain.TreeListSummary(ASender: TcxCustomTreeList; const Arguments: TcxTreeListSummaryEventArguments;
  var OutArguments: TcxTreeListSummaryEventOutArguments);
begin
  if Arguments.Node.IsGroupNode then
    OutArguments.Done := True;
end;

procedure TfrmMain.ValidateExpressionColumn(Sender: TcxTreeListColumn;
  ANode: TcxTreeListNode; const AValue: Variant; AData: TcxEditValidateInfo);
var
  AErrorCode: Integer;
begin
  AErrorCode := ANode.ErrorCodes[Sender.ItemIndex];
  if AErrorCode > 0 then
  begin
    AData.ErrorType := eetError;
    AData.ErrorText := dxSpreadSheetErrorCodeToString(TdxSpreadSheetFormulaErrorCode(AErrorCode));
  end;
end;

procedure TfrmMain.btnShowExpressionEditorClick(Sender: TObject);
begin
  if galColumns.Gallery.Groups[0].Items[0].Checked then
    tlcTotal.ShowExpressionEditor
  else
    tlcAmountDiscount.ShowExpressionEditor;
end;

end.

